IDE error: Access is allowed from Event Dispatch Thread (EDT) only

### Steps

Unsure

### Result

IDE error

### Expected Result

No error

### IDE Details

```
IntelliJ IDEA 2023.1 EAP (Community Edition)
Build #IC-231.7665.28, built on February 23, 2023
Runtime version: 17.0.6+10-b829.1 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 11 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 16
Registry:
    analyze.exceptions.on.the.fly=true

Non-Bundled Plugins:
    com.jetbrains.plugins.ini4idea (231.7665.20)
    it.czerwinski.intellij.wavefront (1.3.0)
    com.intellij.ideolog (203.0.30.0)
    GLSL (1.24)
    org.jetbrains.idea.grammar (2022.3.1)
    com.github.camork.fileExpander (2.1)
    GrepConsole (12.15.211.6693.1)
    org.jetbrains.compose.desktop.ide (1.3.0)

Kotlin: 231-1.8.0-release-345-IJ7665.28
```

### Stack Trace

```
java.util.concurrent.ExecutionException: com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments: Access is allowed from Event Dispatch Thread (EDT) only; see https://jb.gg/ij-platform-threading for details
Current thread: Thread[ApplicationImpl pooled thread 534,4,main] 825934555 (EventQueue.isDispatchThread()=false)
SystemEventQueueThread: Thread[AWT-EventQueue-0,6,main] 532636070
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.lambda$createBackgroundTask$10(BackgroundTaskUtil.java:293)
	at com.intellij.openapi.util.ObjectTree.runWithTrace(ObjectTree.java:127)
	at com.intellij.openapi.util.ObjectTree.executeAll(ObjectTree.java:159)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:264)
	at com.intellij.openapi.util.Disposer.dispose(Disposer.java:252)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.lambda$createBackgroundTask$11(BackgroundTaskUtil.java:315)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1810)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: com.intellij.openapi.diagnostic.RuntimeExceptionWithAttachments: Access is allowed from Event Dispatch Thread (EDT) only; see https://jb.gg/ij-platform-threading for details
Current thread: Thread[ApplicationImpl pooled thread 534,4,main] 825934555 (EventQueue.isDispatchThread()=false)
SystemEventQueueThread: Thread[AWT-EventQueue-0,6,main] 532636070
	at com.intellij.openapi.application.impl.ApplicationImpl.throwThreadAccessException(ApplicationImpl.java:1046)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertIsDispatchThread(ApplicationImpl.java:1033)
	at com.intellij.openapi.editor.impl.EditorImpl.assertIsDispatchThread(EditorImpl.java:3338)
	at com.intellij.openapi.editor.impl.CaretImpl.assertIsDispatchThread(CaretImpl.java:610)
	at com.intellij.openapi.editor.impl.CaretImpl.moveToOffset(CaretImpl.java:99)
	at com.intellij.openapi.editor.CaretModel.moveToOffset(CaretModel.java:91)
	at com.intellij.openapi.editor.CaretModel.moveToOffset(CaretModel.java:77)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialComponent$1.contentsChanged(MtlMaterialComponent.kt:110)
	at java.desktop/javax.swing.AbstractListModel.fireContentsChanged(AbstractListModel.java:127)
	at it.czerwinski.intellij.wavefront.editor.ui.MaterialComboBoxModel.updateMaterials(MaterialComboBoxModel.kt:62)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialComponent$updateMtlFile$$inlined$runReadAction$1.compute(actions.kt:61)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:878)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialComponent.updateMtlFile(MtlMaterialComponent.kt:210)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialComponent.initializeMtlFile(MtlMaterialComponent.kt:150)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialComponent.initialize(MtlMaterialComponent.kt:135)
	at it.czerwinski.intellij.wavefront.editor.MtlMaterialEditor.initPreview$lambda$0(MtlMaterialEditor.kt:130)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.util.BackgroundTaskUtil.lambda$submitTask$8(BackgroundTaskUtil.java:271)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	... 7 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@
 ### Fixed
 - Fix solid shading for OBJ files without texture coordinates
 - Fix rendering of material shading preview for OBJ files without material
+- Fix two-way navigation between MTL text editor and material preview
 
 ### Security
 

--- a/src/main/kotlin/it/czerwinski/intellij/wavefront/editor/MtlMaterialComponent.kt
+++ b/src/main/kotlin/it/czerwinski/intellij/wavefront/editor/MtlMaterialComponent.kt
@@ -17,6 +17,7 @@
 package it.czerwinski.intellij.wavefront.editor
 
 import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.invokeLater
 import com.intellij.openapi.application.runReadAction
 import com.intellij.openapi.editor.event.CaretEvent
 import com.intellij.openapi.editor.event.CaretListener
@@ -107,7 +108,9 @@ class MtlMaterialComponent(
                     val caretModel = SplitEditor.KEY_CARET_MODEL[editor]
                     val materialAtCaret = myMaterialComboBoxModel.findItemAtOffset(caretModel.offset)
                     if (selectedMaterial != null && selectedMaterial != materialAtCaret) {
-                        caretModel.moveToOffset(selectedMaterial.startOffset)
+                        invokeLater {
+                            caretModel.moveToOffset(selectedMaterial.startOffset)
+                        }
                     }
 
                     myMaterialPreviewComponent.updateMaterial(selectedMaterial)
@@ -153,7 +156,11 @@ class MtlMaterialComponent(
     private fun updateMtlFile(mtlFile: MtlFile?) {
         runReadAction {
             myMtlFile = mtlFile
-            myMaterialComboBoxModel.updateMaterials(mtlFile?.materials.orEmpty())
+            myMaterialComboBoxModel.updateMaterials(
+                materials = mtlFile?.materials.orEmpty(),
+                caretOffset = SplitEditor.KEY_CARET_MODEL[editor].offset
+            )
+            myMaterialPreviewComponent.updateMaterial(material)
         }
     }
 

--- a/src/main/kotlin/it/czerwinski/intellij/wavefront/editor/ui/MaterialComboBoxModel.kt
+++ b/src/main/kotlin/it/czerwinski/intellij/wavefront/editor/ui/MaterialComboBoxModel.kt
@@ -51,15 +51,29 @@ class MaterialComboBoxModel(
 
     override fun getSelectedItem(): Any? = mySelected
 
-    fun updateMaterials(materials: Iterable<MtlMaterialElement>) {
+    fun updateMaterials(materials: Iterable<MtlMaterialElement>, caretOffset: Int) {
         if (myList != materials) {
             val oldSize = myList.size
-            myList.clear()
-            myList.addAll(materials)
-            if (mySelected !in myList) {
-                mySelected = myList.firstOrNull()
-            }
+            updateList(materials)
+            updateSelection(caretOffset)
             fireContentsChanged(this, 0, max(oldSize, myList.size))
+        } else if (myList.map(MtlMaterialElement::getName) != materials.map(MtlMaterialElement::getName)) {
+            fireContentsChanged(this, 0, myList.size)
+        }
+    }
+
+    private fun updateList(materials: Iterable<MtlMaterialElement>) {
+        myList.clear()
+        myList.addAll(materials)
+    }
+
+    private fun updateSelection(caretOffset: Int) {
+        val itemAtOffset = findItemAtOffset(caretOffset)
+        if (itemAtOffset != null) {
+            mySelected = itemAtOffset
+        }
+        if (mySelected !in myList) {
+            mySelected = myList.firstOrNull()
         }
     }
 }

